File: daily_python/2975_Maximum_Square_Area_by_Removing_Fences_From_a_Field.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int maximizeSquareArea(int m, int n, vector<int>& hFences, vector<int>& vFences) {
        hFences.push_back(m);
        vFences.push_back(n);
        unordered_set<int> h;
        for (int i = 0; i < hFences.size(); i++){
            h.insert(hFences[i] - 1);
            for (int j = 0; j < i; j++){
                h.insert(abs(hFences[i] - hFences[j]));
            }
        }
        // for (int i : h) cout << i << " ";
        // cout << endl;
        unordered_set<int> v;
        for (int i = 0; i < vFences.size(); i++){
            v.insert(vFences[i] - 1);
            for (int j = 0; j < i; j++){
                v.insert(abs(vFences[i] - vFences[j]));
            }
        }
        // for (int i : v) cout << i << " ";
        // cout << endl;
        int res = -1;
        for (int i : h) if (v.find(i) != v.end()) res = max(res, i);
        if (res == -1) return -1;
        long long ans = (1LL * res * res) % (int(1e9 + 7));
        return ans;
    }
};

int main(){
    int m = 4;
    int n = 3;
    vector<int> hFences = {2,3};
    vector<int> vFences = {2};
    // int m = 6;
    // int n = 7;
    // vector<int> hFences = {2};
    // vector<int> vFences = {4};
    // int m = 3;
    // int n = 9;
    // vector<int> hFences = {2};
    // vector<int> vFences = {8,6,5,4};
    Solution S;
    cout << S.maximizeSquareArea(m, n, hFences, vFences) << endl;
    return 0;
}

File: daily_python/2975_Maximum_Square_Area_by_Removing_Fences_From_a_Field.py
import os
import sys
from typing import List

class Solution:
    def maximizeSquareArea(self, m: int, n: int, hFences: List[int], vFences: List[int]) -> int:
        hFences = sorted([1, m] + hFences)
        vFences = sorted([1, n] + vFences)
        h = { hFences[i] - hFences[j] for i in range(len(hFences)) for j in range(i) }
        v = { vFences[i] - vFences[j] for i in range(len(vFences)) for j in range(i) }
        res = 0
        for d in h:
            if d in v:
                res = max(res, d)
        return -1 if res == 0 else res * res % (10 ** 9 + 7)

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            m = int(lines[0].split('=')[1].strip()[:-1])
            n = int(lines[1].split('=')[1].strip()[:-1])
            hFences = list(map(int, lines[2].split('=')[1].strip()[1:-2].split(',')))
            vFences = list(map(int, lines[3].split('=')[1].strip()[1:-2].split(',')))
            dataset.append((m, n, hFences, vFences))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for m, n, hFences, vFences in dataset:
        results.append(solution.maximizeSquareArea(m, n, hFences, vFences))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')
